File: ios/Message/model/GroupDB.m
//
//  GroupDB.m
//  leveldb_ios
//
//  Created by houxh on 14-7-5.
//  Copyright (c) 2014年 potato. All rights reserved.
//

#import "GroupDB.h"
#import "LevelDB.h"

@implementation GroupDB

+(GroupDB*)instance {
  static GroupDB *db;
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    if (!db) {
      db = [[GroupDB alloc] init];
    }
  });
  return db;
}

-(NSString*)groupKeyFromGroupID:(int64_t)groupID {
    NSString *key = [NSString stringWithFormat:@"groups_%lld", groupID];
    return key;
}

-(BOOL)addGroup:(Group*)group {
    LevelDB *db = [LevelDB defaultLevelDB];
    NSString *key = [self groupKeyFromGroupID:group.groupID];
    
    NSString *k1 = [key stringByAppendingString:@"_topic"];
    NSString *k2 = [key stringByAppendingString:@"_master"];
    NSString *k3 = [key stringByAppendingString:@"_disbanded"];
    
    [db setString:group.topic forKey:k1];
    [db setInt:group.masterID forKey:k2];
    [db setInt:group.disbanded forKey:k3];
    
    for (NSNumber *uid in group.members) {
        [self addGroupMember:group.groupID member:[uid longLongValue]];
    }
    return YES;
}

-(BOOL)removeGroup:(int64_t)groupID {
  LevelDB *db = [LevelDB defaultLevelDB];
  NSString *key = [self groupKeyFromGroupID:groupID];
  
  Group *group = [self loadGroup:groupID];
  if (!group) {
    return NO;
  }
  
  NSString *k1 = [key stringByAppendingString:@"_topic"];
  NSString *k2 = [key stringByAppendingString:@"_master"];

  [db removeValueForKey:k1];
  [db removeValueForKey:k2];
  
  
  for (NSNumber *member in group.members) {
    [self removeGroupMember:group.groupID member:[member longLongValue]];
  }
  return YES;
}

-(BOOL)disbandGroup:(int64_t)groupID {
    LevelDB *db = [LevelDB defaultLevelDB];
    NSString *key = [self groupKeyFromGroupID:groupID];
    
    NSString *k3 = [key stringByAppendingString:@"_disbanded"];
    
    [db setInt:1 forKey:k3];
    return YES;
}

-(NSString*)groupMemberKey:(int64_t)groupID member:(int64_t)uid {
  return [NSString stringWithFormat:@"group_member_%lld_%lld", groupID, uid];
}

-(BOOL)addGroupMember:(int64_t)groupID member:(int64_t)uid {
  LevelDB *db = [LevelDB defaultLevelDB];
  NSString *key = [self groupMemberKey:groupID member:uid];
  [db setString:@"1" forKey:key];
  return YES;
}

-(BOOL)removeGroupMember:(int64_t)groupID member:(int64_t)uid {
  LevelDB *db = [LevelDB defaultLevelDB];
  NSString *key = [self groupMemberKey:groupID member:uid];
  [db removeValueForKey:key];
  return YES;
}

-(NSArray*)loadAllGroup {
  LevelDB *db = [LevelDB defaultLevelDB];
  LevelDBIterator *iter = [db newIterator];
  NSString *t = @"groups_";
  
  NSMutableArray *groups = [NSMutableArray array];
  Group *group = nil;
  for ([iter seek:t]; [iter isValid]; [iter next]) {
    NSString *k = [iter key];
    if (![k hasPrefix:@"groups_"]) {
      break;
    }
    if ([k hasSuffix:@"_topic"]) {
      NSRange range = NSMakeRange(7, k.length-7-6);
      int64_t groupID = [[k substringWithRange:range] longLongValue];
      NSString *topic = [iter value];
      if (groupID == 0) {
        continue;
      }
      group.topic = topic;
      if (group.groupID == groupID && group.masterID > 0) {
        [groups addObject:group];
      }
    } else if ([k hasSuffix:@"_master"]) {
      NSRange range = NSMakeRange(7, k.length-7-7);
      int64_t groupID = [[k substringWithRange:range] longLongValue];
      int64_t master = [[iter value] longLongValue];
      if (groupID == 0 || master == 0) {
        continue;
      }
      
      group = [[Group alloc] init];
      group.masterID = master;
      group.groupID = groupID;
    }
  }
  return groups;
}

-(NSMutableArray*)loadGroupMember:(Group*)group {
  LevelDB *db = [LevelDB defaultLevelDB];
  LevelDBIterator *iter = [db newIterator];
  int64_t groupID = group.groupID;
  NSString *t = @"group_member_";
  for ([iter seek:t]; [iter isValid]; [iter next]) {
    NSString *key = [iter key];
    NSArray *ary = [key componentsSeparatedByString:@"_"];
    if ([ary count] != 4) {
      break;
    }
    if (![[ary objectAtIndex:0] isEqualToString:@"group"] ||
        ![[ary objectAtIndex:1] isEqualToString:@"member"]) {
      break;
    }
    int64_t gid = [[ary objectAtIndex:2] longLongValue];
    int64_t uid = [[ary objectAtIndex:3] longLongValue];
    if (gid != groupID) {
      break;
    }
    [group addMember:uid];
  }
  return nil;
}

-(Group*)loadGroup:(int64_t)groupID {
    LevelDB *db = [LevelDB defaultLevelDB];
    NSString *key = [self groupKeyFromGroupID:groupID];
    
    NSString *k1 = [key stringByAppendingString:@"_topic"];
    NSString *k2 = [key stringByAppendingString:@"_master"];
    NSString *k3 = [key stringByAppendingString:@"_disbanded"];
    
    
    Group *group = [[Group alloc] init];
    group.groupID = groupID;
    group.topic = [db stringForKey:k1];
    group.masterID = [db intForKey:k2];
    group.disbanded = [db intForKey:k3];
    if (group.groupID == 0 || group.masterID == 0 || group.topic == nil) {
        return nil;
    }
    
    [self loadGroupMember:group];
    
    return group;
}

-(NSString*)getGroupTopic:(int64_t)groupID {
    LevelDB *db = [LevelDB defaultLevelDB];
    NSString *key = [self groupKeyFromGroupID:groupID];
    
    NSString *k1 = [key stringByAppendingString:@"_topic"];
    return [db stringForKey:k1];
}

-(void)setGroupTopic:(int64_t)groupID topic:(NSString*)topic {
    LevelDB *db = [LevelDB defaultLevelDB];
    NSString *key = [self groupKeyFromGroupID:groupID];
    
    NSString *k1 = [key stringByAppendingString:@"_topic"];
    
    [db setString:topic forKey:k1];
}

@end

File: android/app/src/main/java/com/beetle/bauhinia/model/GroupDB.java
package com.beetle.bauhinia.model;

import com.google.code.p.leveldb.LevelDB;
import com.google.code.p.leveldb.LevelDBIterator;

import java.lang.reflect.Array;
import java.util.ArrayList;

/**
 * Created by houxh on 15/3/21.
 */
public class GroupDB {

    private static GroupDB instance = new GroupDB();
    public static GroupDB getInstance() {
        return instance;
    }

    private String topicKey(long groupID) {
        return String.format("groups_%d_topic", groupID);
    }

    private String masterKey(long groupID) {
        return String.format("groups_%d_master", groupID);
    }

    private String disbandedKey(long groupID) {
        return String.format("groups_%d_disbanded", groupID);
    }

    private String leavedKey(long groupID) {
        return String.format("groups_%d_leaved", groupID);
    }

    private String groupMemberKey(long groupID, long uid) {
        return String.format("group_member_%d_%d", groupID, uid);
    }

    public boolean addGroupMember(long groupID, long uid)  {
        String key = groupMemberKey(groupID, uid);
        LevelDB db = LevelDB.getDefaultDB();
        try {
            db.set(key, "1");
            return true;
        } catch (Exception e){
            return false;
        }
    }

    public boolean removeGroupMember(long groupID, long uid) {
        String key = groupMemberKey(groupID, uid);
        LevelDB db = LevelDB.getDefaultDB();
        try {
            db.delete(key);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public boolean addGroup(Group group) {
        LevelDB db = LevelDB.getDefaultDB();

        String k1 = topicKey(group.groupID);
        String k2 = masterKey(group.groupID);
        String k3 = disbandedKey(group.groupID);
        try {
            db.set(k1, group.topic);
            db.setLong(k2, group.master);
            db.setLong(k3, group.disbanded ? 1 : 0);

            ArrayList<Long> members = group.getMembers();
            for (Long member: members) {
                addGroupMember(group.groupID, member);
            }
            return true;
        } catch (Exception e) {
            return false;
        }
    }
    public boolean removeGroup(long groupID) {
        LevelDB db = LevelDB.getDefaultDB();

        String k1 = topicKey(groupID);
        String k2 = masterKey(groupID);
        String k3 = disbandedKey(groupID);
        try {
            db.delete(k1);
            db.delete(k2);
            db.delete(k3);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    private ArrayList<Long> loadGroupMember(long groupID) throws Exception {

        LevelDB db = LevelDB.getDefaultDB();
        LevelDBIterator iter = db.newIterator();

        ArrayList<Long> members = new ArrayList<Long>();
        String target = String.format("group_member_%d_", groupID);
        for (iter.seek(target); iter.isValid(); iter.next()) {
            String key = iter.getKey();
            String value = iter.getValue();

            String[] array = key.split("_");
            if (array.length != 4) {
                break;
            }

            if (Long.parseLong(array[2]) != groupID) {
                break;
            }

            Long uid = Long.parseLong(array[3]);

            members.add(uid);
        }

        return members;

    }
    public Group loadGroup(long groupID) {
        LevelDB db = LevelDB.getDefaultDB();
        String k1 = topicKey(groupID);
        String k2 = masterKey(groupID);
        String k3 = disbandedKey(groupID);
        Group group = new Group();

        try {
            group.topic = db.get(k1);
            group.master = db.getLong(k2);
            group.disbanded = db.getLong(k3) == 1 ? true : false;
            group.setMembers(loadGroupMember(groupID));
            return group;
        } catch (Exception e) {
            return null;
        }
    }

    //退出群的标志
    public void leaveGroup(long groupID) {
        LevelDB db = LevelDB.getDefaultDB();

        String k3 = leavedKey(groupID);
        try {
            db.setLong(k3, 1);
        } catch (Exception e) {

        }
    }

    //清空退出群的标志
    public void joinGroup(long groupID) {
        LevelDB db = LevelDB.getDefaultDB();

        String k3 = leavedKey(groupID);
        try {
            db.setLong(k3, 0);
        } catch (Exception e) {

        }
    }

    public boolean isLeaved(long groupID) {
        LevelDB db = LevelDB.getDefaultDB();

        String k3 = leavedKey(groupID);
        try {
            long leaved = db.getLong(k3);
            return (leaved == 1);
        } catch (Exception e) {
            return false;
        }
    }

    public boolean disbandGroup(long groupID) {
        LevelDB db = LevelDB.getDefaultDB();

        String k3 = disbandedKey(groupID);
        try {
            db.setLong(k3, 1);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public String getGroupTopic(long groupID) {
        LevelDB db = LevelDB.getDefaultDB();

        String k3 = topicKey(groupID);
        try {
            return db.get(k3);
        } catch (Exception e) {
            return "";
        }
    }

    public boolean setGroupTopic(long groupID, String name) {
        LevelDB db = LevelDB.getDefaultDB();

        String k3 = topicKey(groupID);
        try {
            db.set(k3, name);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public boolean setGroupMaster(long groupID, long master) {
        LevelDB db = LevelDB.getDefaultDB();

        String k3 = masterKey(groupID);
        try {
            db.setLong(k3, master);
            return true;
        } catch (Exception e) {
            return false;
        }
    }
}
